### PR TITLE
Lab 2 - Sorting

### LINKED ISSUE #181: Lab 2 - Sorting
Lab 2

### PR DESCRIPTION
**Pull request by n0npax**
Fixes #181 
Review of colleague's PR #176

#### Changes proposed in this PR:
3 sort algorithms - Lab02:
- Implement BubbleSort
- Implement Insertion
- Implement QuickSort
- Add Tests

### CODE REVIEW

File: 02_bubble/n0npax/main.go [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+import (
+	"fmt"
+	"io"
+	"os"
+)
+
+var out io.Writer = os.Stdout
+
+// Divides slice into partitions base on pivot hight
+func partition(s []int, low int, high int) int {
```
Comment (juliaogris) [inline 02_bubble/n0npax/main.go:12]: Go function comment convention is to first use function name - this way godocs are generated:
```
// partition Divides slice into partitions based on pivot high
```
same for the functions below.

s/hight/high
s/base/based

File: 02_bubble/n0npax/main.go [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+		quicksort(s, lo, p-1)
+		quicksort(s, p+1, hi)
+	}
+
+}
+
+// Return sorted list(copy) using quicksort
+// Wrapper to unify sort definition
+// Quick sort has average compexity O(nlogn)
+func quicksortWrapper(s []int) []int {
```
Comment (juliaogris) [inline 02_bubble/n0npax/main.go]: I'd call this `quicksort` and the one above `qs`, `qsort` or `quicksortInPlace` or so.

File: 02_bubble/n0npax/main.go [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+		}
+	}
+	s[i+1], s[high] = s[high], s[i+1]
+	return i + 1
+}
+
+// Sorts provided partition. Modifies reference
+func quicksort(s []int, lo int, hi int) {
+	if lo < hi {
+		// p - partition veretical pivot
```
Comment (juliaogris) [inline 02_bubble/n0npax/main.go]: veretical?

File: 02_bubble/n0npax/main.go [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+			if v2 > v1 {
+				ts[i1], ts[i2] = ts[i2], ts[i1]
+			}
+		}
+	}
+	return ts
+}
+
+// Copy slice
+func copySlice(s []int) []int {
```
Comment (juliaogris) [inline 02_bubble/n0npax/main.go]: if you insist on a oneline you could do this: `b = append(a[:0:0], a...)` (https://github.com/go101/go101/wiki/How-to-perfectly-clone-a-slice%3F)

File: 02_bubble/n0npax/main_test.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+	}
+}
+
+func TestQuicksort(t *testing.T) {
+	for size := 0; size < 10; size++ {
+		testFunction(t, quicksortWrapper, "quicksort", size)
+	}
+}
+
+func testFunction(t *testing.T, f fn, name string, size int) {
```
Comment (juliaogris) [inline 02_bubble/n0npax/main_test.go]: could be `TestSort` with `testCases map[string]fn{ "bubble": bubble, "quick": quicksortWrapper/*...*/ }` than loop over the different sort implementations.

**Review by juliaogris — CHANGES_REQUESTED** [review]
Looks good, a few nits.
Please rename your first commit from `Lab1 ..` to something like `Implement lab 1` see [notes on commit messages](https://github.com/anz-bank/go-course/wiki/Lab-2:-Collected-Code-Review-Comments#commit-messages)

File: 02_bubble/n0npax/main_test.go (hunk 5 above)
Comment (n0npax) [reply 02_bubble/n0npax/main_test.go]: Thanks for that! Fixed

File: 02_bubble/n0npax/main.go (hunk 4 above)
Comment (n0npax) [reply 02_bubble/n0npax/main.go]: fixed. Thanks for hint.

File: 02_bubble/n0npax/main.go (hunk 3 above)
Comment (n0npax) [reply 02_bubble/n0npax/main.go]: fixed

File: 02_bubble/n0npax/main.go (hunk 2 above)
Comment (n0npax) [reply 02_bubble/n0npax/main.go]: fixed Thanks!

File: 02_bubble/n0npax/main.go (hunk 1 above)
Comment (n0npax) [reply 02_bubble/n0npax/main.go:12]: fixed. Thanks!

File: 02_bubble/n0npax/main.go (hunk 4 above)
Comment (n0npax) [reply 02_bubble/n0npax/main.go]: I mean I changed function body to one liner. Do you want me to replace all function invocation with provided syntax?

**Comment by n0npax** [comment]
@juliaogris, @rokane  could you please review changes ?

**Comment by n0npax** [comment]
@juliaogris @rokane Could you review this PR please?

File: 02_bubble/n0npax/main_test.go [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+
+	expected := strconv.Quote("[1 2 3 5]\n")
+	actual := strconv.Quote(buf.String())
+
+	if expected != actual {
+		t.Errorf("Unexpected output. Expected: %q - Actual: %q", expected, actual)
+	}
+}
+
+func randomSlice(size int) []int {
```
Comment (juliaogris) [inline 02_bubble/n0npax/main_test.go:25]: nice!

File: 02_bubble/n0npax/main.go [hunk 7]
```diff
@@ -0,0 +1,75 @@
...
+		}
+	}
+	return ts
+}
+
+// bubble sorts list(copy) using buble sort
+func bubble(s []int) []int {
+	ts := append(s[:0:0], s...)
+
+	for i1, v1 := range ts {
```
Comment (juliaogris) [inline 02_bubble/n0npax/main.go]: Sorry I should have picked up on this early as well, but my understanding of bubble sort is:
- first iteration: *pairwise* comparison and swapping from `0` to `len-1`. 
   This ensures that the largest element is place in `ts[len-1]`
- second iteration: *pairwise* comparison and swapping from `0` to `len-2`. 
  This ensures that the second larges element is placed in `ts[len-2]`
...

So one by one, the largest, second larges and so forth elements "bubble" up to the top.

Here's our "canonical" solution after a bit of bikeshedding:
```go
	for i := len(s); i > 0; i-- {
		for j := 1; j < i; j++ {
			if s[j-1] > s[j] {
				s[j-1], s[j] = s[j], s[j-1]
			}
		}
	}
```

Here with early break (if nothing's been swapped the slice is sorted):
```go
	for i := len(s); i > 0; i-- {
		swapped := false
		for j := 1; j < i; j++ {
			if s[j-1] > s[j] {
				s[j-1], s[j] = s[j], s[j-1]
				swapped = true
			}
		}
		if !swapped {
			// If no elements were swapped, the slice is sorted so we can stop here
			return
		}
	}
```

File: 02_bubble/n0npax/main.go [hunk 8]
```diff
@@ -0,0 +1,75 @@
...
+// insertion sorts list(copy) using insertion
+func insertion(s []int) []int {
+	ts := append(s[:0:0], s...)
+
+	for i := range ts {
+		for i > 0 {
+			if ts[i-1] > ts[i] {
+				ts[i-1], ts[i] = ts[i], ts[i-1]
+			}
+			i--
```
Comment (juliaogris) [inline 02_bubble/n0npax/main.go:52]: Should have commented on this earlier (and no need to change anything).
I've learned by now that the you have implemented `insertion` is a generally accepted solution. 
For completeness, the way I've learned it and Donald Knuth describes it in insertion sort you build up a sorted slice from left to right by finding the right "insertion point" for each each successive element in the already sorted slice on the left.
It's a bit like how you might sort a deck of cards in your hand: you go from left to right, pull out one card at a time and place it in the correct position in the already sorted part.

Here's the solution that we've come up with (after some programming golf):
```go
	for i := 1; i < len(s); i++ {
		var j int
		for j = i; j > 0; j-- {
			if s[j-1] <= s[i] {
				break // We've found the insertion point
			}
		}
		// insert s[i] at s[j] moving the rest up
		tmp := s[i]
		copy(s[j+1:i+1], s[j:i])
		s[j] = tmp
	}
```

**Review by juliaogris — DISMISSED** [review]
Looks good - however, I think bubble sort should be adjusted a little to deserve its name. Feel free to rebut. I've posted our canonical solutions that we've come up with after a bit of debate - feel free to use, adjust or ignore :)

File: 02_bubble/n0npax/main.go (hunk 7 above)
Comment (n0npax) [reply 02_bubble/n0npax/main.go]: Fixed @juliaogris

File: 02_bubble/n0npax/main_test.go [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+	testCase := map[string]fn{
+		"bubble":    bubble,
+		"insertion": insertion,
+		"quicksort": quicksort,
+	}
+	for size := 0; size <= 10; size++ {
+		for name, testFunc := range testCase {
+			testFunc, size := testFunc, size
+			t.Run(name, func(t *testing.T) {
+				s := randomSlice(size)
```
Comment (rokane) [inline 02_bubble/n0npax/main_test.go:45]: Nice work!

**Review by rokane — APPROVED** [review]
Looks good :) nice work implementing the quicksort algorithm as well!

**Pull request merged**